Detectors/TRD/raw/src/EpnDataReaderTask.cxx

```diff
@@ -0,0 +1,97 @@
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// @file   EpnDataReaderTask.cxx
+/// @author Sean Murray
+/// @brief  TRD cru output to tracklet task
+
+#include "TRDRaw/EpnDataReaderTask.h"
+#include "TRDRaw/CruRawReader.h"
+#include "Framework/ControlService.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/RawDeviceService.h"
+#include "Framework/DeviceSpec.h"
+#include "Framework/DataSpecUtils.h"
+
+#include <fairmq/FairMQDevice.h>
+
+using namespace o2::framework;
+
+namespace o2
+{
+namespace trd
+{
+
+void EpnDataReaderTask::init(InitContext& ic)
+{
+  LOG(INFO) << "Cru2Tracklet Task init";
+
+  auto finishFunction = [this]() {
+    mReader.checkSummary();
+  };
+
+  ic.services().get<CallbackService>().set(CallbackService::Id::Stop, finishFunction);
+}
+
+void EpnDataReaderTask::run(ProcessingContext& pc)
+{
+  LOG(info) << "TRD Translator Task run";
+
+  /* set encoder output buffer */
+  char bufferOut[1048576];
+
+  auto device = pc.services().get<o2::framework::RawDeviceService>().device();
+  auto outputRoutes = pc.services().get<o2::framework::RawDeviceService>().spec().outputs;
+  auto fairMQChannel = outputRoutes.at(0).channel;
+  int inputcount = 0;
+  /* loop over inputs routes */
+  for (auto iit = pc.inputs().begin(), iend = pc.inputs().end(); iit != iend; ++iit) {
+    if (!iit.isValid())
+      continue;
+    //LOG(info) << "iit.mInputs  " << iit.mInputs.
+    /* prepare output parts */
+    FairMQParts parts;
+
+    /* loop over input parts */
+    for (auto const& ref : iit) {
+
+      auto headerIn = DataRefUtils::getHeader<o2::header::DataHeader*>(ref);
+      auto dataProcessingHeaderIn = DataRefUtils::getHeader<o2::framework::DataProcessingHeader*>(ref);
+      auto payloadIn = ref.payload;
+      auto payloadInSize = headerIn->payloadSize;
+      mReader.setDataBuffer(payloadIn);
+      mReader.setDataBufferSize(payloadInSize);
+      /* run */
+      mReader.run();
+      auto payloadOutSize = 1; //mReader.getEncoderByteCounter();
+      auto payloadMessage = device->NewMessage(payloadOutSize);
+      std::memcpy(payloadMessage->GetData(), bufferOut, payloadOutSize);
+
+      /* output */
+      auto headerOut = *headerIn;
+      auto dataProcessingHeaderOut = *dataProcessingHeaderIn;
+      headerOut.dataDescription = "TRDRAW";
+      headerOut.payloadSize = payloadOutSize;
+      o2::header::Stack headerStack{headerOut, dataProcessingHeaderOut};
+      auto headerMessage = device->NewMessage(headerStack.size());
+      std::memcpy(headerMessage->GetData(), headerStack.data(), headerStack.size());
+
+      /* add parts */
+      parts.AddPart(std::move(headerMessage));
+      parts.AddPart(std::move(payloadMessage));
+    }
+
+    /* send message */
+    device->Send(parts, fairMQChannel);
```

Comment:
I don't understand inputs and input parts, but let me comment anyways.

It looks like you send one message per input, where you concatenate all the parts. If one input is e.g. "digits" and the parts are the 12 FLPs, then this looks ok - as far as I can tell with my non-existent knowledge.

But I would still expect (maybe) to have one message with everything (tracklets, digits... for all FLPs) going out. But maybe it should actually be separate output messages...

Reply:
I basically implemented both tasks in the same way. I can trivially restructure the epndatareader more like the other ones in trd.